Blog post about first steps on SCM integration

saraycp opened the pull request:
Link to the blog post: https://deploy-preview-294--openbuildservice.netlify.app/2021/05/31/scm-integration/

hennevogel commented:
Deploy preview for *openbuildservice* ready!

Built with commit 9c28da2f022518d5426b71fd98c8675f56ac3a96

https://deploy-preview-294--openbuildservice.netlify.app

dmarcoux commented on _posts/development/2021-05-04-scm-integration.md:
```diff
@@ -0,0 +1,142 @@
+
+Most of the code implemented nowadays is managed collaboratively
+through source code management (SCM) systems like GitHub, GitLab or Pagure.
+More and more people want to handle their package sources with those systems too,
```
```suggestion
An increasing number of people want to handle their package sources with those systems too,
```

dmarcoux commented on _posts/development/2021-05-04-scm-integration.md:
```diff
@@ -0,0 +1,142 @@
+Most of the code implemented nowadays is managed collaboratively
+through source code management (SCM) systems like GitHub, GitLab or Pagure.
+More and more people want to handle their package sources with those systems too,
+what leads us to detect a lack of better integration between SCMs and OBS.
```
```suggestion
and while there is an integration between SCMs and OBS, it could be definitely improved.
```

dmarcoux commented on _posts/development/2021-05-04-scm-integration.md:
```diff
@@ -0,0 +1,142 @@
+
+### 1. Set a Configuration File
+
+Add the configuration file `.obs/workflows.yml` in the root directory of your GitHub or GitLab repository.
```
I would be extra clear here about the file, so even telling people to create the file in the `.obs` folder.

```suggestion
In the root directory of your GitHub/GitLab repository, create a directory `.obs` which contains the configuration file `workflows.yml`.
```

dmarcoux commented on _posts/development/2021-05-04-scm-integration.md:
```diff
@@ -0,0 +1,142 @@
+
+#### Set a GitHub Webhook
+
+In GitHub you can go to *Settings -> Webhooks*.
```
```suggestion
In GitHub, you can go to *Settings -> Webhooks*.
```

dmarcoux commented on _posts/development/2021-05-04-scm-integration.md:
```diff
@@ -0,0 +1,142 @@
+
+#### Set a GitLab Webhook
+
+In GitLab you can go to *Settings -> Webhooks*.
```
```suggestion
In GitLab, you can go to *Settings -> Webhooks*.
```

dmarcoux commented on _posts/development/2021-05-04-scm-integration.md:
```diff
@@ -0,0 +1,142 @@
+
+As soon as you add the webhook, OBS will start receiving requests for every Merge Request happening in GitLab.
+
+### 5. The Workflow is Triggered in OBS
```
```suggestion
### 5. The Workflow Is Triggered in OBS
```

dmarcoux commented on _posts/development/2021-05-04-scm-integration.md:
```diff
@@ -0,0 +1,142 @@
+---
+layout: post
+title: Heading for an SCMs Integration in OBS
```
```suggestion
title: Heading for a Better SCM Integration in OBS
```

adrianschroeter commented:
On Mittwoch, 5. Mai 2021, 11:03:46 CEST Dany Marcoux wrote:
> @dmarcoux commented on this pull request.
> 
> 
> 
> > @@ -0,0 +1,142 @@
> +---
> +layout: post
> +title: Heading for an SCMs Integration in OBS
> 
> ```suggestion
> title: Heading for a Better SCM Integration in OBS

this is not about SCM integration, it stays untouched.

It is about integrating external submission workflows via pull/merge requests.

-- 

Adrian Schroeter ***@***.***>
Build Infrastructure Project Manager

SUSE Software Solutions Germany GmbH,  Maxfeldstr. 5, 90409 Nuernberg, Germany
(HRB 247165, AG München), Geschäftsführer: Felix Imendörffer

dmarcoux commented:
@adrianschroeter It is about SCM integration, you ever wrote it yourself: `It is about integrating external submission workflows via pull/merge requests.`. This is one of the use cases, there are more which will be developed in the future.

adrianschroeter commented:
On Mittwoch, 5. Mai 2021, 11:23:32 CEST Dany Marcoux wrote:
> @adrianschroeter It is about SCM integration, you ever wrote it yourself: `It is about integrating external submission workflows via pull/merge requests.`. This is one of the use cases, there are more which will be developed in the future.

well, I am just worried about a mixup with SCM integration which exist and
is about come.
(like existing scm tooling and upcoming external git repositories for projects)

The topic is wide indeed, my feeling is that we should be a bit more
specific when talking about the concrete functionality.

Something like SCM review workflow integration with external platforms like
gitlab/github.

-- 

Adrian Schroeter ***@***.***>
Build Infrastructure Project Manager

SUSE Software Solutions Germany GmbH,  Maxfeldstr. 5, 90409 Nuernberg, Germany
(HRB 247165, AG München), Geschäftsführer: Felix Imendörffer

dmarcoux commented:
This is literally what the blog post is about... We do explain that this is for GitHub and GitLab. We cannot summarize everything in the title. Do you have a suggestion for the blog post title?

adrianschroeter commented:
On Mittwoch, 5. Mai 2021, 11:43:59 CEST Dany Marcoux wrote:
> This is literally what the blog post is about... We do explain that this is for GitHub and GitLab. We cannot summarize everything in the title. Do you have a suggestion for the blog post title?

what about "GitLab and GitHub workflow integration" ?

-- 

Adrian Schroeter ***@***.***>
Build Infrastructure Project Manager

SUSE Software Solutions Germany GmbH,  Maxfeldstr. 5, 90409 Nuernberg, Germany
(HRB 247165, AG München), Geschäftsführer: Felix Imendörffer

saraycp commented:
Thanks @dmarcoux for your suggestions, they are addressed now.
I finally wrote an hybrid title: `Heading for a GitHub and GitLab Workflow Integration in OBS`. Hope you find it better, @adrianschroeter

adrianschroeter commented:
On Mittwoch, 5. Mai 2021, 14:27:32 CEST Saray Cabrera Padrón wrote:
> Thanks @dmarcoux for your suggestions, they are addressed now.
> I finally wrote an hybrid title: `Heading for a GitHub and GitLab Workflow Integration in OBS`. Hope you find it better, @adrianschroeter 

sounds good :)

-- 

Adrian Schroeter ***@***.***>
Build Infrastructure Project Manager

SUSE Software Solutions Germany GmbH,  Maxfeldstr. 5, 90409 Nuernberg, Germany
(HRB 247165, AG München), Geschäftsführer: Felix Imendörffer

eduardoj commented on _posts/development/2021-05-04-scm-integration.md:
````diff
@@ -0,0 +1,143 @@
+        destination_package: ctris
+```
+
+For now, OBS only accepts one workflow with a single step: `branch_package`.
````
Nitpick: I would explain here that the destination project will be the concatenation of the value of `destination_project_prefix` and the number of the pull request as suffix.

saraycp replied:
I rephrased the sentence to clarify that.

I also removed the `filter` part on the .yml I use like an example.

hennevogel commented on _posts/development/2021-05-04-scm-integration.md:
```diff
@@ -0,0 +1,135 @@
+most expected one among OBS users: managing the package sources in an SCM system
+and use OBS to build the source code and report back the results to GitHub or GitLab.
+
+We propose the following steps to achieve this goal. Give them a try!
```
I think you should talk a little bit more about which process you want to make possible with those changes.

- every PR/MR is triggering a package branch on OBS
- OBS is reporting back the build status of the branched package to the PR/MR

Then dive into the steps that people can do to try this. I know you have described this below but it's way better to tell people up-front what this is about.

saraycp replied:
I added two sentences paragraph the whole idea: https://github.com/openSUSE/obs-landing/pull/294/commits/7bf243757183aee3beaac07b62a3dc9a1a169174#diff-643f2999b778582f99173bf8c79448e41f5e21eb147322f94705781cf890c564R19-R21

hennevogel commented on _posts/development/2021-05-04-scm-integration.md:
```diff
@@ -0,0 +1,135 @@
+
+Apart from the repository inside an SCM system, you have to create an analogous project/package inside OBS.
+
+For example, for a repository called `ctris` in GitHub/GitLab you might have a `home:Iggy/ctris` project and package in OBS.
```
This makes it sound as if those names have anything to do with this. But you can have a package called `hans` on OBS and a repository called `franz` on your SCM. There is no relation by name.

Way more important than the name is the whole integration with [obs-service-tar_scm](https://github.com/openSUSE/obs-service-tar_scm) and it's ability to fetch the sources from the PR/MR source fork/branch.

saraycp replied:
Now I use different names.

The explanation about the integration with `obs-service-tar_scm` is still pending.

hennevogel commented on _posts/development/2021-05-04-scm-integration.md:
```diff
@@ -0,0 +1,135 @@
+
+Adding the `scm_token` part is not really needed to generate the token. It is just an extra value containing the GitHub/GitLab token
+that OBS have to use to report back to the SCM system.
+
```
You are missing the step before this one to create personal access tokens on the SCM.

saraycp replied:
It was explained later. But I moved the explanation to the second step, so it's clear from the beginning.

hennevogel commented on _posts/development/2021-05-04-scm-integration.md:
````diff
@@ -0,0 +1,135 @@
+
+```
+osc token --create --operation workflow scm_token=ssx_j2du52Ozs8oZAtxq3K0bdy9TEp8iUy4JilTs
+```
````
Please warn people about the consequences of leaking the OBS token secrect.

saraycp replied:
Now there is a WARNING paragraph :)
https://github.com/openSUSE/obs-landing/pull/294/commits/7bf243757183aee3beaac07b62a3dc9a1a169174#diff-643f2999b778582f99173bf8c79448e41f5e21eb147322f94705781cf890c564R76-R77

hennevogel commented on _posts/development/2021-05-04-scm-integration.md:
```diff
@@ -0,0 +1,135 @@
+  <img src="/images/posts/sprint_95_checks_gitlab.png" alt="Checks GitLab" />
+  <figcaption>Checks GitLab</figcaption>
+</figure>
+
```
You should also add a paragraph about all the limitations of this implementation that we already know.

- we don't handle PR syncs like new commits/force push etc. yet
- we only report back build results for one repository/arch (right?)
- what happens if I manually click rebuild on OBS?
- more?

And additionally maybe you want to give an outlook on what's more to come?

- token UI
- branch/event filtering
- more types of steps
- reporting for whole projects with many repositories
- maybe even mention outgoing webhooks

Just to get people excited for this journey. This is the first time they hear about this.

hennevogel commented on _posts/development/2021-05-04-scm-integration.md:
```diff
@@ -0,0 +1,168 @@
+You have to fill in the form with
+- *URL*: the OBS endpoint to trigger the workflow operation (https://build.opensuse.org/trigger/workflow?id=0). Ensure you replace the `id` with the one obtained in the token creation (step 4).
+- *Secret Token*: the OBS's token string obtained in step 4.
+- *Trigger*. We just need you to select `Merge Requests`.
```
The sreenshot doesn't reflect this

saraycp replied:
Screenshot updated, now only "Merge Requests" is selected.

hennevogel commented on _posts/development/2021-05-31-scm-integration.md:
```diff
@@ -0,0 +1,171 @@
+You have to fill in the form with
+- *URL*: the OBS endpoint to trigger the workflow operation (https://build.opensuse.org/trigger/workflow?id=0). Ensure you replace the `id` with the one obtained in the token creation (step 4).
+- *Secret Token*: the OBS's token string obtained in step 4.
+- *Trigger*. We just need you to select `Merge Requests`.
```
The screenshot doesn't reflect this.

saraycp replied:
Screenshot updated, now only "Merge Requests" is selected.

saraycp commented on _posts/development/2021-05-31-scm-integration.md:
```diff
@@ -0,0 +1,171 @@
+
+We have started to integrate GitHub and GitLab for a scenario that seems to be the
+most expected one among OBS users: managing the package sources in an SCM system
+and use OBS to build the source code and report back the results to GitHub or GitLab.
```
Done

saraycp commented on _posts/development/2021-05-31-scm-integration.md:
```diff
@@ -0,0 +1,171 @@
+will trigger a package branch in OBS. Then, as soon as the branched package's building finishes,
+OBS will report back the build status to the pull/merge request.
+
+We propose the following steps to achieve this goal. Give them a try!
```
Done

saraycp commented on _posts/development/2021-05-31-scm-integration.md:
```diff
@@ -0,0 +1,171 @@
+
+Most of the code implemented nowadays is managed collaboratively
+through source code management (SCM) systems like GitHub, GitLab or Pagure.
+An increasing number of people want to handle their package sources with those systems too
+and while there is an integration between SCMs and OBS, it could definitely be improved.
```
As `An increasing number of you want to...` sounds weird to me I've decided to write `Every day more of you want to... `

dmarcoux replied:
I'm fine with that change. I don't have a better suggestion.

saraycp commented on _posts/development/2021-05-31-scm-integration.md:
```diff
@@ -0,0 +1,171 @@
+For now, OBS only accepts one workflow with a single step: `branch_package`.
+Adapt the names of the source project and package to your needs.
+
+In the example above, we went with the workflow name `workflow1`. You can be more creative!
```
This whole section has been merged with the previous one and the whole text has changed.

saraycp commented on _posts/development/2021-05-31-scm-integration.md:
```diff
@@ -0,0 +1,171 @@
+
+### 2. Create Project and Package in OBS
+
+Apart from the repository inside an SCM system, you have to create an analogous project/package inside OBS.
```
Rephrased.

The `_service` stuff is still a TODO.

saraycp commented on _posts/development/2021-05-31-scm-integration.md:
```diff
@@ -0,0 +1,171 @@
+
+In the example above, we went with the workflow name `workflow1`. You can be more creative!
+
+The step `branch_package` indicates to OBS that you want to branch the project/package given by `source_project` and `source_package` fields.
```
This whole section has been merged with the previous one and the whole text has changed.

saraycp commented on _posts/development/2021-05-31-scm-integration.md:
````diff
@@ -0,0 +1,171 @@
+        source_package: ctris
+```
+
+For now, OBS only accepts one workflow with a single step: `branch_package`.
````
This whole section has been merged with the previous one and the whole text has changed.

saraycp commented on _posts/development/2021-05-31-scm-integration.md:
````diff
@@ -0,0 +1,171 @@
+```
+
+For now, OBS only accepts one workflow with a single step: `branch_package`.
+Adapt the names of the source project and package to your needs.
````
This whole section has been merged with the previous one and the whole text has changed.

saraycp commented:
@dmarcoux I squashed your commit with the existing one. Then I addressed @hennevogel 's suggestions and finally re-structured the text a bit.

I still have some work to do, but you can tell me if the little change in the structure is enough for you.
Now there are two main sections to describe the workflow: `Workflow Integration Steps` (containing the numeric steps) and `Workflow Triggering and Back Reporting` (explaining what happens after the configuration is completed).

adrianschroeter commented on _posts/development/2021-05-31-scm-integration.md:
````diff
@@ -0,0 +1,159 @@
+```
+
+This indicates that, on every new pull/merge request in the repository's side, OBS will branch the project/package given by
+`source_project` and `source_package` fields and will try to build them.
````
Shouldn't it be enough to register the token for the package? Why do we need an additional file with the same information?

adrianschroeter commented on _posts/development/2021-05-31-scm-integration.md:
````diff
@@ -0,0 +1,159 @@
+
+```
+osc token --create --operation workflow scm_token=ssx_j2du52Ozs8oZAtxq3K0bdy9TEp8iUy4JilTs
+```
````
it would be good to have the example for a specific package here. The global tokens (permitting it in any package where the user has maintainer rights) should not be used by default. 

It would mean to give write access to the SCM hoster for any package, maybe osc should not even allow this by default...

hennevogel replied:
The workflow tokens are never bound to a package. They are just bound to a users and what they can and can't do.

dmarcoux replied:
> It would mean to give write access to the SCM hoster for any package, maybe osc should not even allow this by default...

This is why people shouldn't leak their SCM token. Same for the OBS token, we also warn users about this in this blog post.

krauselukas commented on _posts/development/2021-05-31-scm-integration.md:
```diff
@@ -0,0 +1,176 @@
+`source_project` and `source_package` fields and will try to build them.
+
+For all of this to be possible, the OBS workflows depend on the [obs-service-tar_scm](https://github.com/openSUSE/obs-service-tar_scm) service which is configured in the `_service` file of the source package.
+This service has to be configutred to point to your GitHub/GitLab repository.
```
```suggestion
This service has to be configured to point to your GitHub/GitLab repository.
```
```suggestion
This service has to be configutred to point to your GitHub/GitLab repository.
```

krauselukas commented on _posts/development/2021-05-31-scm-integration.md:
````diff
@@ -0,0 +1,176 @@
+</service>
+```
+
+Without a service, we cannot pull the source changes in OBS, thus rendering OBS workflows pretty much useless.
````
@dmarcoux I wouldn't go into much details regarding the `_branch_request` file, but just thinking that it might be worth to mention that OBS will take care of the rest (like pointing to the right revision/commit once the workflow got triggered)

krauselukas commented on _posts/development/2021-05-31-scm-integration.md:
```diff
@@ -0,0 +1,176 @@
+through source code management (SCM) systems like GitHub, GitLab or Pagure.
+An increasing number of you want to handle their package sources with those systems too.
+And while there is an integration between SCMs and OBS, it could definitely be extended and improved.
+In today's post, we are sharing how we began to move to meet that need.
```
Hmm, I think we are more showing what is possible and how to use it

krauselukas replied:
```suggestion
In today's post, we are sharing what's possible in these regards from now on and how to use it.```

saraycp commented on _posts/development/2021-05-31-scm-integration.md:
```diff
@@ -0,0 +1,183 @@
+1. OBS executes the workflow you configured
+1. OBS reports the status (pending, success or failure) of the workflow to the pull/merge request that was opened
+
+There are some prerequisites we assume you already have already. First, if you haven't done it already, you have to join [the beta program](https://openbuildservice.org/2018/10/04/the-beta-program/), so you can try this new feature. And surprise surprise, you will need a repository on Github/Gitlab and a package on build.opensuse.org you want to integrate! 🤓
```
```suggestion
There are some prerequisites we assume you already have. First, if you haven't done it yet, you have to join [the beta program](https://openbuildservice.org/2018/10/04/the-beta-program/), so you can try this new feature. And surprise surprise, you will need a repository on Github/Gitlab and a package on build.opensuse.org you want to integrate! 🤓
```

saraycp commented on _posts/development/2021-05-31-scm-integration.md:
```diff
@@ -0,0 +1,183 @@
+
+## Authentication Setup
+
+First of all we have to make sure that OBS and your SCM service are *allowed* to talk to each other. Tokens are the
```
```suggestion
First of all, we have to make sure that OBS and your SCM service are *allowed* to talk to each other. Tokens are the
```

saraycp commented on _posts/development/2021-05-31-scm-integration.md:
```diff
@@ -0,0 +1,183 @@
+
+You have to give OBS permissions to report the build status to the pull/merge request to the SCM on your behalf.
+
+For this you have to create a personal access token. Check [GitHub's](https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token) and [GitLab's](https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html#creating-a-personal-access-token) documentation to learn how.
```
```suggestion
For this, you have to create a personal access token. Check [GitHub's](https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token) and [GitLab's](https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html#creating-a-personal-access-token) documentation to learn how.
```

saraycp commented on _posts/development/2021-05-31-scm-integration.md:
```diff
@@ -0,0 +1,183 @@
+
+For this you have to create a personal access token. Check [GitHub's](https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token) and [GitLab's](https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html#creating-a-personal-access-token) documentation to learn how.
+
+Make sure the access token you generate has the *at least* those scopes assigned:
```
```suggestion
Make sure the access token you generate has *at least* those scopes assigned:
```

saraycp commented on _posts/development/2021-05-31-scm-integration.md:
````diff
@@ -0,0 +1,183 @@
+</status>
+```
+
+Of course you should provide your real SCM personal access token here, created in the previous step.
````
```suggestion
Of course, you should provide your real SCM personal access token here, created in the previous step.
```

saraycp commented on _posts/development/2021-05-31-scm-integration.md:
```diff
@@ -0,0 +1,183 @@
+
+Of course you should provide your real SCM personal access token here, created in the previous step.
+By specifying the operation `workflow` you limit the scope of this token to this kind of operation.
+Note down this access token and it's ID, you are going to use them in the next step. For the sake of
```
```suggestion
Note down this access token and its ID, you are going to use them in the next step. For the sake of
```

saraycp commented on _posts/development/2021-05-31-scm-integration.md:
```diff
@@ -0,0 +1,183 @@
+Of course you should provide your real SCM personal access token here, created in the previous step.
+By specifying the operation `workflow` you limit the scope of this token to this kind of operation.
+Note down this access token and it's ID, you are going to use them in the next step. For the sake of
+clarity, let's pretend your OBS access token is: `uvwxyz` and it's numerical ID is `12345`.
```
```suggestion
clarity, let's pretend your OBS access token is: `uvwxyz` and its numerical ID is `12345`.
```

saraycp commented on _posts/development/2021-05-31-scm-integration.md:
```diff
@@ -0,0 +1,183 @@
+By the way, there are also various other types of tokens in the OBS that can help you to authenticate for
+operations in OBS. Check out [the manual](https://openbuildservice.org/help/manuals/obs-user-guide/cha.obs.authorization.token.html#id-1.5.10.16.4) if you're interested.
+
+## Setup a Webhook
```
```suggestion
## Set up a Webhook
```

saraycp commented on _posts/development/2021-05-31-scm-integration.md:
```diff
@@ -0,0 +1,183 @@
+  <br /> Replace `uvwxyz` with the OBS token secret string obtained above.
+- **Trigger**: `Merge Requests`
+
+## Setup the OBS Workflow
```
```suggestion
## Set up the OBS Workflow
```

saraycp commented on _posts/development/2021-05-31-scm-integration.md:
```diff
@@ -0,0 +1,183 @@
+
+Now that the two services talk to each other, we need to configure what should happen on OBS when a new
+pull/merge request is opened. To do so, in the root directory of your GitHub/GitLab repository, create a directory
+`.obs` which contains the a file called `workflows.yml`.
```
```suggestion
`.obs` which contains a file called `workflows.yml`.
```

saraycp commented on _posts/development/2021-05-31-scm-integration.md:
````diff
@@ -0,0 +1,183 @@
+        source_package: ctris
+```
+
+- **source_project**: Configure the name of the project that containts the package you want to test build
````
```suggestion
- **source_project**: Configure the name of the project that contains the package you want to test build
```

saraycp commented on _posts/development/2021-05-31-scm-integration.md:
````diff
@@ -0,0 +1,183 @@
+```
+
+- **source_project**: Configure the name of the project that containts the package you want to test build
+- **source_package**: Confoigure the name of the package you want to test build
````
```suggestion
- **source_package**: Configure the name of the package you want to test build
```

saraycp commented on _posts/development/2021-05-31-scm-integration.md:
```diff
@@ -0,0 +1,183 @@
+created. What happens if a new commit is added to the pull/merge request or if someone force pushes some changes?
+We don't handle this yet.
+
+Depending on the package you branch, it might build for more than one repository or architecture. Currently
```
```suggestion
Depending on the package your branch, it might build for more than one repository or architecture. Currently
```

dmarcoux replied:
The version with `you` is correct.

danidoni commented on _posts/development/2021-05-31-scm-integration.md:
```diff
@@ -0,0 +1,186 @@
+---
+
+Most of the code you implement nowadays is managed collaboratively through source code management (SCM) systems
+like GitHub, GitLab or Pagure. And an increasing number of you also want to handle their package sources with those systems.
```
```suggestion
like GitHub, GitLab or Pagure. An increasing number of you also want to handle their package sources with those systems.
```

danidoni replied:
And maybe join that sentence with the previous one

dmarcoux replied:
I find it good as it is.

danidoni replied:
Okay!

danidoni commented on _posts/development/2021-05-31-scm-integration.md:
```diff
@@ -0,0 +1,186 @@
+
+Depending on the package you branch, it might build for more than one repository or architecture. Currently
+we will report the *last* build result that occurs.
+
+[//]: # FIXME: is that so?
+We also don't report back the result of a manually triggered rebuild in OBS.
```
```suggestion
```

saraycp commented:
LGTM! :+1:

adrianschroeter commented:
On Donnerstag, 27. Mai 2021, 11:15:31 CEST Dany Marcoux wrote:
> @dmarcoux commented on this pull request.
> 
> 
> 
> > +to learn how to create a personal access token.
> +
> +In the next step, you will learn how to pass the so-called `scm_token` to OBS, this token represents your user account on the SCM system's side.
> +
> +#### 3. Create a Workflow Token in OBS
> +
> +As you want to trigger some actions in OBS as a response to the events hapenning GitHub/GitLab,
> +you also need an OBS's access token. GitHub/GitLab will use it to trigger the actions on your behalf.
> +
> +The token represents your user account and the kind of action/operation you want to perform, which is `workflow` in this case.
> +
> +You can create a token in OBS using `osc`:
> +
> +```
> +osc token --create --operation workflow scm_token=ssx_j2du52Ozs8oZAtxq3K0bdy9TEp8iUy4JilTs
> +```
> 
> > It would mean to give write access to the SCM hoster for any package, maybe osc should not even allow this by default...
> 
> This is why people shouldn't leak their SCM token. Same for their OBS tokens, we also warn users about this in this blog post.

sure, but a leaked token (or when the remote service is doing invalid things) should only affect the object and operation 
bound to the token. This is the hole concept of security tokens as we have it.

Therefore the limitation to that object and operation must happen on OBS side.

Otherwise you would need again to create own user accounts for each operation to achieve the same.


-- 

Adrian Schroeter ***@***.***>
Build Infrastructure Project Manager

SUSE Software Solutions Germany GmbH,  Maxfeldstr. 5, 90409 Nuernberg, Germany
(HRB 247165, AG München), Geschäftsführer: Felix Imendörffer

dmarcoux commented:
Regarding tokens and security, we didn't change anything in regards to how the `runservice`, `release` and `rebuild` tokens act. They can be assigned to a specific package and in this case, they can act upon this package. If they aren't assigned to a package, then it's whatever the user is allowed to do.

As for the new `workflow` token, if it is somehow leaked, all workflow steps act solely on the home project of the user linked to the token. This is how we manage risks.
